### PR TITLE
use middleware to capture HTTP server metrics

### PR DESCRIPTION
**Pull request by libotony**
# Description

Using middleware to capture HTTP server metrics, added names for all paths, added feature that only records named path's metrics to avoid random API access potentially blowing up metrics db.

In this PR, **dependency upgraded** `github.com/gorilla/mux` v1.6.0 -> v1.8.1

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] New and existing E2E tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have not added any vulnerable dependencies to my code

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -40,14 +40,15 @@ func New(
...
 	allowedOrigins string,
 	backtraceLimit uint32,
 	callGasLimit uint64,
 	pprofOn bool,
 	skipLogs bool,
 	allowCustomTracer bool,
-	isReqLoggerEnabled bool,
-	forkConfig thor.ForkConfig,
+	enableReqLogger bool,
+	enableMetrics bool,
```
Comment (otherview) [inline api/api.go:51]: `isReqLoggerEnabled` is a boolean bc it's a question. But I understand not everyone's' cup of tea.

Perhaps the past tense ? `enabledReqLogger` and `enabledMetrics`?

File: api/accounts/accounts.go [hunk 2]
```diff
@@ -354,21 +354,27 @@ func (a *Accounts) Mount(root *mux.Router, pathPrefix string) {
 
 	sub.Path("/*").
 		Methods(http.MethodPost).
-		HandlerFunc(utils.MetricsWrapHandlerFunc(pathPrefix, "accounts_call_batch_code", a.handleCallBatchCode))
+		Name("accounts_call_batch_code").
+		HandlerFunc(utils.WrapHandlerFunc(a.handleCallBatchCode))
 	sub.Path("/{address}").
 		Methods(http.MethodGet).
-		HandlerFunc(utils.MetricsWrapHandlerFunc(pathPrefix, "accounts_get_account", a.handleGetAccount))
+		Name("accounts_get_detail").
```
Comment (otherview) [inline api/accounts/accounts.go:361]: Would be good to keep the name as close as possible to the URL and avoid the interpretation step.
Ie. `curl .../accounts/0x0000` -> `accounts_get_account`

File: api/accounts/accounts.go [hunk 3]
```diff
@@ -354,21 +354,27 @@ func (a *Accounts) Mount(root *mux.Router, pathPrefix string) {
...
+		HandlerFunc(utils.WrapHandlerFunc(a.handleGetCode))
 	sub.Path("/{address}/storage/{key}").
 		Methods("GET").
-		HandlerFunc(utils.MetricsWrapHandlerFunc(pathPrefix, "accounts_get_storage", a.handleGetStorage))
+		Name("accounts_get_storage").
+		HandlerFunc(utils.WrapHandlerFunc(a.handleGetStorage))
 	// These two methods are currently deprecated
 	sub.Path("").
 		Methods(http.MethodPost).
-		HandlerFunc(utils.MetricsWrapHandlerFunc(pathPrefix, "accounts_api_call_contract", a.handleCallContract))
```
Comment (otherview) [inline api/accounts/accounts.go:370]: good spot 👍

File: api/blocks/blocks.go [hunk 4]
```diff
@@ -95,5 +95,6 @@ func (b *Blocks) Mount(root *mux.Router, pathPrefix string) {
 	sub := root.PathPrefix(pathPrefix).Subrouter()
 	sub.Path("/{revision}").
 		Methods(http.MethodGet).
-		HandlerFunc(utils.MetricsWrapHandlerFunc(pathPrefix, "blocks_get_block", b.handleGetBlock))
+		Name("blocks_get_detail").
```
Comment (otherview) [inline api/blocks/blocks.go:98]: Would be good to keep the name as close as possible to the URL and avoid the interpretation step.
Ie. `curl .../block/0x0000` -> `blocks_get_block`

File: api/events/events.go [hunk 5]
```diff
@@ -63,5 +63,6 @@ func (e *Events) Mount(root *mux.Router, pathPrefix string) {
 
 	sub.Path("").
 		Methods(http.MethodPost).
-		HandlerFunc(utils.MetricsWrapHandlerFunc(pathPrefix, "events_filter", e.handleFilter))
+		Name("events_filter").
```
Comment (otherview) [inline api/events/events.go:66]: Would be good to keep the name as close as possible to the URL and avoid the interpretation step.
Ie. `curl .../logs/event` -> `logs_get_event`

File: api/metrics.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+		var (
+			enabled = false
+			name    = ""
+		)
+
+		// all named route will be recorded
+		if rt != nil && rt.GetName() != "" {
+			enabled = true
+			name = rt.GetName()
+		}
```
Comment (otherview) [inline api/metrics.go:51]: Good stuff 👍 I think this works well.

Would be good to have some unit tests around these edge cases where the route is nil or the route not having a name.

File: api/metrics.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+		now := time.Now()
+		mrw := newMetricsResponseWriter(w)
+
+		next.ServeHTTP(mrw, r)
+
+		if enabled {
+			metricHttpReqCounter().AddWithLabel(1, map[string]string{"name": name, "code": strconv.Itoa(mrw.statusCode), "method": r.Method})
+			metricHttpReqDuration().ObserveWithLabels(time.Since(now).Milliseconds(), map[string]string{"name": name, "code": strconv.Itoa(mrw.statusCode), "method": r.Method})
```
Comment (otherview) [inline api/metrics.go:60]: Just to tripple check, are we sure this captures errors ?

File: api/node/node.go [hunk 8]
```diff
@@ -35,5 +35,6 @@ func (n *Node) Mount(root *mux.Router, pathPrefix string) {
 
 	sub.Path("/network/peers").
 		Methods(http.MethodGet).
-		HandlerFunc(utils.MetricsWrapHandlerFunc(pathPrefix, "node_network", n.handleNetwork))
+		Name("node_network_peers").
+		HandlerFunc(utils.WrapHandlerFunc(n.handleNetwork))
```
Comment (otherview) [inline api/node/node.go:39]: Would be good to keep the name as close as possible to the URL and avoid the interpretation step.
Ie. `curl .../network/peers` -> `network_get_peers`

File: api/transactions/transactions.go [hunk 9]
```diff
@@ -219,11 +219,14 @@ func (t *Transactions) Mount(root *mux.Router, pathPrefix string) {
 
 	sub.Path("").
 		Methods(http.MethodPost).
-		HandlerFunc(utils.MetricsWrapHandlerFunc(pathPrefix, "transactions_send_transaction", t.handleSendTransaction))
+		Name("transactions_send").
```
Comment (otherview) [inline api/transactions/transactions.go:222]: Suggestion:
`transactions_send_tx`

File: api/transactions/transactions.go [hunk 10]
```diff
@@ -219,11 +219,14 @@ func (t *Transactions) Mount(root *mux.Router, pathPrefix string) {
...
+		HandlerFunc(utils.WrapHandlerFunc(t.handleSendTransaction))
 	sub.Path("/{id}").
 		Methods(http.MethodGet).
-		HandlerFunc(utils.MetricsWrapHandlerFunc(pathPrefix, "transactions_get_transaction_by_id", t.handleGetTransactionByID))
+		Name("transactions_get_detail").
```
Comment (otherview) [inline api/transactions/transactions.go:226]: Suggestion:
`transactions_get_tx`

File: api/transfers/transfers.go [hunk 11]
```diff
@@ -69,5 +69,6 @@ func (t *Transfers) Mount(root *mux.Router, pathPrefix string) {
 
 	sub.Path("").
 		Methods(http.MethodPost).
-		HandlerFunc(utils.MetricsWrapHandlerFunc(pathPrefix, "transfers_transfer_logs", t.handleFilterTransferLogs))
+		Name("transfers_filter").
```
Comment (otherview) [inline api/transfers/transfers.go:72]: Suggestion:
`logs_get_transfer`

File: api/api.go (hunk 1 above)
Comment (libotony) [reply api/api.go:51]: For past tense, `reqLoggerEnabled` and `metricsEnabled`, the current one just followed the flag's name convention.

File: cmd/thor/flags.go [hunk 12]
```diff
@@ -127,7 +127,7 @@ var (
 		Name:  "disable-pruner",
 		Usage: "disable state pruner to keep all history",
 	}
-	metricsEnabledFlag = cli.BoolFlag{
+	enableMetricsFlag = cli.BoolFlag{
```
[the remark this answers is not in the document]
Comment (libotony) [reply cmd/thor/flags.go:130]: Done in the metrics PR.

File: api/node/node.go (hunk 8 above)
Comment (libotony) [reply api/node/node.go:39]: I would go for `node_get_network_peers` since it is under the path `node`

File: api/metrics.go (hunk 7 above)
Comment (libotony) [reply api/metrics.go:60]: Yes, test case also added

**Comment by libotony** [comment]
@otherview @leszek-vechain Unit test added, covering both 200 and none-200 cases and also no named urls(`metrics`) and scenario that `rt=nil`

File: api/api.go (hunk 1 above)
Comment (otherview) [reply api/api.go:51]: Yeah that's true, sounds good 👍

**Review by otherview — APPROVED** [review]
